ui/app/components/clients/line-chart.ts

```diff
@@ -0,0 +1,145 @@
+/**
+ * Copyright (c) HashiCorp, Inc.
+ * SPDX-License-Identifier: BUSL-1.1
+ */
+
+import Component from '@glimmer/component';
+import { SVG_DIMENSIONS, formatNumbers } from 'vault/utils/chart-helpers';
+import { parseAPITimestamp } from 'core/utils/date-formatters';
+import { format, isValid, parse } from 'date-fns';
+import { SerializedChartData, UpgradeData } from 'vault/client-counts';
+import { debug } from '@ember/debug';
+
+interface Args {
+  dataset: SerializedChartData[];
+  upgradeData: UpgradeData[];
+  xKey?: string;
+  yKey?: string;
+  chartHeight?: number;
+  dateFormat?: string;
+}
+
+interface ChartData {
+  x: Date;
+  y: number | null;
+  new: number;
+  tooltipUpgrade: string | null;
+}
+
+interface UpgradeByMonth {
+  [key: string]: UpgradeData;
+}
+
+/**
+ * @module LineChart
+ * LineChart components are used to display time-based data in a line plot with accompanying tooltip
+ *
+ * @example
+ * ```js
+ * <LineChart @dataset={{dataset}} @upgradeData={{this.versionHistory}}/>
+ * ```
+ * @param {string} xKey - string denoting key for x-axis data of dataset. Should reference a date string with format 'M/yy'.
+ * @param {string} yKey - string denoting key for y-axis data of dataset. Should reference a number or null.
+ * @param {array} upgradeData - array of objects containing version history from the /version-history endpoint
+ * @param {number} [chartHeight=190] - height of chart in pixels
+ * @param {string} [dateFormat='M/yy'] - format of date string in xKey
+ */
+export default class LineChart extends Component<Args> {
+  // Chart settings
+  get yKey() {
+    return this.args.yKey || 'clients';
+  }
+  get xKey() {
+    return this.args.xKey || 'month';
+  }
+  get chartHeight() {
+    return this.args.chartHeight || SVG_DIMENSIONS.height;
+  }
+  get dateFormat() {
+    return this.args.dateFormat || 'M/yy';
+  }
+  // Plot points
+  get data(): ChartData[] {
+    try {
+      return this.args.dataset?.map((datum) => {
+        const timestamp = parse(datum[this.xKey] as string, this.dateFormat, new Date()) as Date;
+        if (isValid(timestamp) === false)
+          throw new Error(
+            `Unable to parse value "${datum[this.xKey]}" as date with format ${this.dateFormat}`
+          );
+        const upgradeMessage = this.getUpgradeMessage(datum);
+        return {
+          month: datum[this.xKey],
+          x: timestamp,
```

Comment:
Does this have to be a date object? I'd expect the existing `datum.timestamp` which is an ISO string to work? Then we don't have to introduce localized timezones which happens when calling `new Date()`